File: Baekjoon/Others/13902.cpp
#include<iostream>
#include<algorithm> // fill
using namespace std;
#define INF 987654321
#define min(x, y) (x > y) ? y : x
int memo[10001];
int wok[1000];
bool amount[10001];

int main() {
	// 0 입력
	int N, M, i, j, k;
	cin >> N >> M;
	for (i = 0; i < M; i++)
		cin >> wok[i];

	// 1 하나 또는 두개의 웍을 사용해서 만들수 있는 요리의 양 계산
	for (i = 0; i < M; i++)
	{
		amount[wok[i]] = true;
		for (j = i + 1; j < M; j++) {
			if (wok[i] + wok[j] > N) continue;
			amount[wok[i] + wok[j]] = true;
		}
	}
	// 2 dp계산
	fill(memo, (memo + N + 1), INF);
	memo[0] = 0;
	for (i = 0; i <= N; i++)
		if (amount[i]) memo[i] = 1;
	for (i = 0; i <= N; i++)
	{
		for (k = 0; k <= N; k++)
		{
			if (!amount[k]) continue;
			if (i + k > N) continue;
			memo[i + k] = min(memo[i] + 1, memo[i + k]);
		}
	}
	// 3 출력
	if (memo[N] == INF)
		cout << -1 << '\n';
	else
		cout << memo[N] << '\n';
	return 0;
}

File: Baekjoon/Others/13902.py
# import sys; input = sys.stdin.readline
N, M = map(int, input().split())
wok = list(map(int, input().split()))
wok_set = [False for _ in range(10001)]
for i in range(M):
    wok_set[wok[i]] = True
    for j in range(i + 1, M):
        if wok[i] + wok[j] > N: continue
        wok_set[wok[i] + wok[j]] = True
INF = 1234567890
memo = [INF for _ in range(10001)]
memo[0] = 0
for i in range(N + 1):
    if wok_set[i]:
        memo[i] = 1;
for i in range(N):
    for k in range(N):
        if not wok_set[k]: continue
        if i + k > N: continue
        memo[i + k] = min(memo[i] + 1, memo[i + k])

if INF == memo[N]:
    print(-1)
else:
    print(memo[N])
